net: improve the interface around FindNode() and avoid a recursive mutex lock

vasild opened the pull request:
`CConnman::FindNode()` would lock `m_nodes_mutex`, find the node in `m_nodes`, release the mutex and return the node. The current code is safe but it is a dangerous interface where a caller may end up using the node returned from `FindNode()` without owning `m_nodes_mutex` and without having that node's reference count incremented.

Change `FindNode()` to return a boolean since all but one of its callers used its return value to check whether a node exists and did not do anything else with the return value.

Remove a recursive lock on `m_nodes_mutex`.

Rename `FindNode()` to better describe what it does.

vasild committed:
net: merge AlreadyConnectedToAddress() and FindNode(CNetAddr)

`CConnman::AlreadyConnectedToAddress()` is the only caller of
`CConnman::FindNode(CNetAddr)`, so merge the two in one function.

The unit test that checked whether `AlreadyConnectedToAddress()` ignores
the port is now unnecessary because now the function takes a `CNetAddr`
argument. It has no access to the port.

vasild committed:
net: avoid recursive m_nodes_mutex lock in DisconnectNode()

Have `CConnman::DisconnectNode()` iterate `m_nodes` itself instead of
using `FindNode()`. This avoids recursive mutex lock and drops the only
caller of `FindNode()` which used the return value for something else
than a boolean found/notfound.

vasild committed:
net: change FindNode() to not return a node and rename it

All callers of `CConnman::FindNode()` use its return value `CNode*` only
as a boolean null/notnull. So change that method to return `bool`.

This removes the dangerous pattern of handling a `CNode` object (the
return value of `FindNode()`) without holding `CConnman::m_nodes_mutex`
and without having that object's reference count incremented for the
duration of the usage.

Also rename the method to better describe what it does.

vasild commented on src/net.cpp:
```diff
@@ -404,10 +389,8 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
-        CNode* pnode = FindNode(static_cast<CService>(addrConnect));
-        if (pnode)
-        {
-            LogPrintf("Failed to open new connection, already connected\n");
+        if (OutboundConnectedToService(addrConnect)) {
```
This rename makes obvious that maybe something is wrong in the current code in `master`. Is the intention really to ignore inbound connections when, according to the comment we intent to "Look for an existing connection"?

For example, if `1.1.1.1` connects to `2.2.2.2:8333`, then the TCP connection will look like:
`1.1.1.1:53481 <-> 2.2.2.2:8333` where `53481` is a random port picked by the initiator's OS. Now, if `2.2.2.2` intends to connect to `1.1.1.1:8333` then this check will return false because `1.1.1.1:53481` is not `1.1.1.1:8333` and the code will proceed with connecting.

Is this really the intention? Shouldn't this check ignore the port?

jonatack replied:
I ought to resuscitate some of https://github.com/bitcoin/bitcoin/pull/28248 that was looking at this (or will review if you do).

mzumsande replied:
I don't know if this is intentional, or a bug or something in betweeen.

If I remember correctly , #28248 suggested to change that, or something very similar. 
But changing it is a bit scary to review, because sometimes IP addresses are shared - especially local IPs such as 127.0.0.1. So just changing it to ignore the port would have the potential to make local setups that probably exist out there not work any longer.

jonatack replied:
Yes, I see review feedback on that in https://github.com/bitcoin/bitcoin/pull/28248#issuecomment-1789702441 and https://github.com/bitcoin/bitcoin/pull/28248#discussion_r1370747413.

vasild replied:
Love how you deciphered this at an instant! :heart:

I will look into adding a comment to the call sites about this.

vasild replied:
> I will look into adding a comment to the call sites about this.

Can't come up with a good comment. Leaving it as it is, unless somebody has suggestions.

vasild replied:
> I assume addr-gossip only contains address+port from _outbound_ peers ...

The origin of the gossip is the self-announcement done by each node:

https://github.com/bitcoin/bitcoin/blob/458720e5e98c6e9103aea1fdfcd39bafc26c27bb/src/net_processing.cpp#L5286-L5300

then as this address is received by a node, it is further relayed to other nodes. Eventually everybody stores it in their addrman. Then [`GETADDR` requests](https://developer.bitcoin.org/reference/p2p_networking.html#getaddr) are served from addrman. Outbound connections are made to addresses that are already in addrman.

hodlinator replied:
Good that we advertise our listening port for outbound peers! Seems kind of unnecessary to tell inbound peers about the address they report as seeing us as (`CService{node.GetAddrLocal()}`):
https://github.com/bitcoin/bitcoin/blob/458720e5e98c6e9103aea1fdfcd39bafc26c27bb/src/net.cpp#L246-L266
But I guess it simplifies the addrman-logic a little bit, not having to populate/refresh it from the local outbound connections.

theuni replied:
> Hmm, in `CConnman::OpenNetworkConnection()` the code does:
> 
>     * if `pszDest` is not set then use `AlreadyConnectedToAddress(addrConnect)` which ignores the port
> 
>     * if `pszDest` is set then `OutboundConnectedToStr(pszDest)` which compares the port as well.
> 
> 
> Changing that logic is, I guess, out of the scope of this non-functional refactor PR. If to be done it would better be assessed in isolation in its own PR.

Also, later on in `ConnectNode()`:

```c++
    if (pszDest == nullptr) {
        if (IsLocal(addrConnect))
            return nullptr;

        // Look for an existing connection
        CNode* pnode = FindNode(static_cast<CService>(addrConnect));
        if (pnode)
        {
            LogPrintf("Failed to open new connection, already connected\n");
            return nullptr;
        }
```

Which seems to be redundant, as it's a strictly narrower check because of the port?

vasild replied:
True, and `ConnectNode()` is only called from `OpenNetworkConnection()`. Looks like another room for improvement.

While this may seem minor, the amount of developers' time spent on grasping it is not. In the past I have stared at these calls disproportionally long. Would be nice if this is spared to future developers. I will go after that separately from this PR.

jonatack commented:
Review ACK 1e415d2cb703fcd1b622e5343fa264425601c8d modulo CI fixup

Happy to see these methods see an update.

I tried adding `EXCLUSIVE_LOCKS_REQUIRED(!m_nodes_mutex)` to the declarations, but it looks like that lock is already held by some of their callers.

vasild commented on src/net.cpp:
```diff
@@ -332,42 +332,27 @@ bool IsLocal(const CService& addr)
 bool CConnman::AlreadyConnectedToAddress(const CAddress& addr)
 {
-    return FindNode(static_cast<CNetAddr>(addr)) || FindNode(addr.ToStringAddrPort());
+    const CNetAddr net_addr{addr};
```
Fixed, thanks!

jonatack commented:
ACK cc8f239663a0874c307efa30815b40d9ef40badb

w0xlt commented on src/net.cpp:
```diff
@@ -332,42 +332,27 @@ bool IsLocal(const CService& addr)
 }
 
-CNode* CConnman::FindNode(const CNetAddr& ip)
+bool CConnman::OutboundConnectedToStr(const std::string& str_addr)
```
`FindNode()` is not bad, but I personally like `ConnectedToAddrPort()`. This name makes the purpose of the function very clear.
The parameter type (`CService` and `str`) makes the difference clear. It is not necessary for the function name to describe the parameter.

hodlinator replied:
These are the best I could come up with:
* `IsConnectedToStrAddr`
* {`IsConnectedToService` |  `IsConnectedToAddrPort`}
* `IsConnectedToAddress`

`Is`-prefix == short way to signify boolean query function.

vasild replied:
Done.

hodlinator reviewed (approved):
ACK cc8f239663a0874c307efa30815b40d9ef40badb

Fixes thread-unsafe pattern (originally?) raised by me here: https://github.com/bitcoin/bitcoin/pull/32015#discussion_r1988127685
(Even though current master is not susceptible to the found `CNode` being deleted randomly on another thread).

shahsb reviewed (commented):
Concept ACK

hodlinator commented on src/net.cpp:
```diff
@@ -332,42 +332,27 @@ bool IsLocal(const CService& addr)
 bool CConnman::AlreadyConnectedToAddress(const CAddress& addr)
 {
-    return FindNode(static_cast<CNetAddr>(addr)) || FindNode(addr.ToStringAddrPort());
+    const CNetAddr& net_addr{addr};
```
Inconsistency seems to have been introduced when we started checking for string-version in 9bab521df895c149579b9e64931405c56b008afb in April of 2012. Not on purpose I would wager.

hodlinator replied:
Nice!

theuni commented:
Is making `m_nodes_mutex` non-recursive a goal? That sounds nice to me and I'm not opposed, but I'd rather not make changes just to get halfway there.

Concept ACK on turning the `FindNode`s into bools, though. That's a nice improvement.

hebasto commented:
> Is making `m_nodes_mutex` non-recursive a goal? That sounds nice to me and I'm not opposed, but I'd rather not make changes just to get halfway there.

I'm curious how far we are from a non-recursive `m_nodes_mutex`?

vasild commented:
> Is making `m_nodes_mutex` non-recursive a goal?

No, I did not think about that before you mentioned it. The goal here is to make the interface around `FindNode()` safer and simpler (don't return a possibly dangling pointer and don't return a pointer at all since callers can do with a boolean).

I did mention dropping the recursive lock because I think less recursive locking is good, even if the mutex remains recursive.

> I'm curious how far we are from a non-recursive `m_nodes_mutex`?

This far away:

<details>
<summary>[patch] make m_nodes_mutex non-recursive</summary>

```diff
--- i/src/net_processing.cpp
+++ w/src/net_processing.cpp
@@ -1284,29 +1284,29 @@ void PeerManagerImpl::MaybeSetPeerAsAnnouncingHeaderAndIDs(NodeId nodeid)
                 std::swap(lNodesAnnouncingHeaderAndIDs.front(), *std::next(lNodesAnnouncingHeaderAndIDs.begin()));
             }
         }
     }
     m_connman.ForNode(nodeid, [this](CNode* pfrom) EXCLUSIVE_LOCKS_REQUIRED(::cs_main) {
         AssertLockHeld(::cs_main);
-        if (lNodesAnnouncingHeaderAndIDs.size() >= 3) {
-            // As per BIP152, we only get 3 of our peers to announce
-            // blocks using compact encodings.
-            m_connman.ForNode(lNodesAnnouncingHeaderAndIDs.front(), [this](CNode* pnodeStop){
-                MakeAndPushMessage(*pnodeStop, NetMsgType::SENDCMPCT, /*high_bandwidth=*/false, /*version=*/CMPCTBLOCKS_VERSION);
-                // save BIP152 bandwidth state: we select peer to be low-bandwidth
-                pnodeStop->m_bip152_highbandwidth_to = false;
-                return true;
-            });
-            lNodesAnnouncingHeaderAndIDs.pop_front();
-        }
         MakeAndPushMessage(*pfrom, NetMsgType::SENDCMPCT, /*high_bandwidth=*/true, /*version=*/CMPCTBLOCKS_VERSION);
         // save BIP152 bandwidth state: we select peer to be high-bandwidth
         pfrom->m_bip152_highbandwidth_to = true;
         lNodesAnnouncingHeaderAndIDs.push_back(pfrom->GetId());
         return true;
     });
+    if (lNodesAnnouncingHeaderAndIDs.size() > 3) {
+        // As per BIP152, we only get 3 of our peers to announce
+        // blocks using compact encodings.
+        m_connman.ForNode(lNodesAnnouncingHeaderAndIDs.front(), [this](CNode* pnodeStop){
+            MakeAndPushMessage(*pnodeStop, NetMsgType::SENDCMPCT, /*high_bandwidth=*/false, /*version=*/CMPCTBLOCKS_VERSION);
+            // save BIP152 bandwidth state: we select peer to be low-bandwidth
+            pnodeStop->m_bip152_highbandwidth_to = false;
+            return true;
+        });
+        lNodesAnnouncingHeaderAndIDs.pop_front();
+    }
 }
 
 bool PeerManagerImpl::TipMayBeStale()
 {
     AssertLockHeld(cs_main);
     const Consensus::Params& consensusParams = m_chainparams.GetConsensus();
```
</details>

On top of this PR, plus a pile of `EXCLUSIVE_LOCKS_REQUIRED(!m_nodes_mutex)` additions in `net.h`. Should I PR that (separately from the current PR)?

theuni commented:
> I'd appreciate it. It would be good to cross `m_nodes_mutex` off the [list](https://github.com/bitcoin/bitcoin/issues/19303).

Yeah, agreed. That's a good justification for the change to non-recursive here.

Looking at that diff, it's not clear to me if it's safe to move that hunk out of `cs_main`, but we can review/test/argue about it in a new PR.

vasild commented:
Converted to draft, waiting for:

* https://github.com/bitcoin/bitcoin/pull/32338 to make it
* then change this PR to use naming as in https://github.com/bitcoin/bitcoin/pull/32326#discussion_r2063650801

Further improvements away from this PR:
* make `m_nodes_mutex` non-recursive, https://github.com/bitcoin/bitcoin/pull/32326#issuecomment-2835048806
* assess all the checks that compare just the address or address+port and remove redundant ones and make the calls consistent regardless of whether `pzsDest` is set or not, https://github.com/bitcoin/bitcoin/pull/32326#discussion_r2065614037

vasild commented:
`cc8f239663...8ef769cc2d`: rebase and use the naming:

```cpp
bool IsConnectedToAddrPort(const std::string& addr_port);
bool IsConnectedToAddrPort(const CService& addr_port);
bool IsConnectedToAddr(const CNetAddr& addr);
```

vasild commented on src/net.cpp:
```diff
@@ -404,10 +389,8 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
-        {
-            LogPrintf("Failed to open new connection, already connected\n");
+        if (OutboundConnectedToService(addrConnect)) {
+            LogPrintf("Failed to open new connection to %s, already connected", addrConnect.ToStringAddrPort());
```
Done in latest push.

vasild commented:
Made `m_nodes_mutex` non-recursive in https://github.com/bitcoin/bitcoin/pull/32394 which is this PR + one more commit.

hodlinator reviewed (approved):
re-ACK 4f635d100dd1ab5e83bb392d8fbee4a3f4400315

Concept unchanged since [first ACK](https://github.com/bitcoin/bitcoin/pull/32326#pullrequestreview-2790161149).

Changes since then:
* Includes now merged #32338 which simplified `AlreadyConnectedToAddress()` (spawned from this PR: https://github.com/bitcoin/bitcoin/pull/32326#discussion_r2058384454)
* Changed from `OutboundConnectedTo(Str|Service)` -> `IsConnectedToAddr(|Port) const` and `str_addr` -> `addr_port`, nice!
* `AlreadyConnectedToAddress(const CAddress&)` -> `IsConnectedToAddr(const CNetAddr&) const`.

jonatack commented on src/net.h:
```diff
@@ -1352,15 +1352,28 @@ class CConnman
+     * Determine whether we're already connected to a given address.
      */
-    bool AlreadyConnectedToAddress(const CAddress& addr);
+    bool IsConnectedToAddr(const CNetAddr& addr) const;
```
I prefer the original function naming `AlreadyConnectedToAddress` and it looks like retaining it would also reduce the diff (`IsConnectedToAddrPort` could be `AlreadyConnectedToAddressPort` instead).

vasild replied:
That would mean change from:

```cpp
bool IsConnectedToAddrPort(const std::string& addr_port);
bool IsConnectedToAddrPort(const CService& addr_port);
bool IsConnectedToAddr(const CNetAddr& addr);
```

to:

```cpp
bool AlreadyConnectedToAddressPort(const std::string& addr_port);
bool AlreadyConnectedToAddressPort(const CService& addr_port);
bool AlreadyConnectedToAddress(const CNetAddr& addr);
```

I am fine either way. @hodlinator, what do you think?

hodlinator replied:
Previous thread: https://github.com/bitcoin/bitcoin/pull/32326#discussion_r2058445224

`AlreadyConnectedToAddressPort` is significantly more verbose than master's `FindNode`. Checked for uses of `Is[A-Z]\w`-prefix and the `Already`-prefix seems to rarely work, meaning `Already` is more specific in this case. This convinced me that `Already` might be fine. :+1:

vasild replied:
Changed to

```cpp
bool AlreadyConnectedToAddressPort(const std::string& addr_port);
bool AlreadyConnectedToAddressPort(const CService& addr_port);
bool AlreadyConnectedToAddress(const CNetAddr& addr);
```

vasild commented:
`4f635d100d...8668b0c64d`: address https://github.com/bitcoin/bitcoin/pull/32326#discussion_r2072443916

hodlinator reviewed (approved):
re-ACK 8668b0c64d59d4dd7207bf8347b23b8b50aeb7c3

Just changed prefixes from `Is-` -> `Already-` and nicely simplified message in initial commit as per https://github.com/bitcoin/bitcoin/pull/32326#discussion_r2072443916.

vasild committed:
doc: clarify peer address in getpeerinfo and addnode RPC help

The returned value in `getpeerinfo/addr` could be a hostname as well as
an IP address and the `:port` part could be missing. It is displayed
from `CNode::m_addr_name` which could have been set from RPC `addnode`
where the argument is allowed to be a hostname and an optional port.

vasild commented on src/net.cpp:
```diff
@@ -332,26 +332,16 @@ bool IsLocal(const CService& addr)
 }
 
-CNode* CConnman::FindNode(const std::string& addrName)
+bool CConnman::AlreadyConnectedToAddressPort(const std::string& addr_port) const
```
There were previous discussions about the names of these methods: https://github.com/bitcoin/bitcoin/pull/32326#discussion_r2054679774 and https://github.com/bitcoin/bitcoin/pull/32326#discussion_r2072443916

I now renamed the method discussed above, so all 3 are changed from:
```cpp
// previous version of the PR
AlreadyConnectedToAddressPort(const std::string& addr_port);
AlreadyConnectedToAddressPort(const CService& addr_port);
AlreadyConnectedToAddress(const CNetAddr& addr);
```
to
```cpp
// new
AlreadyConnectedToHost(const std::string& host);
AlreadyConnectedToAddressPort(const CService& addr_port);
AlreadyConnectedToAddress(const CNetAddr& addr);
```
this removes the overload, making it more grep-friendly.

Also added one more commit to clarify the RPC help texts.

sipa reviewed (commented):
utACK c9ffb6d7104f016ab89057bd57cd128ac03fd1fc

vasild commented on src/net.cpp:
```diff
@@ -3631,9 +3631,11 @@ void CConnman::GetNodeStats(std::vector<CNodeStats>& vstats) const
-        pnode->fDisconnect = true;
+    auto it = std::ranges::find_if(m_nodes, [&strNode](CNode* node) { return node->m_addr_name == strNode; });
+    if (it != m_nodes.end()) {
+        auto node{*it};
```
Done.

furszy reviewed (commented):
Code review ACK 87e7f37918d42c28033e9f684db52f94eeed617b

hodlinator reviewed (approved):
re-ACK 87e7f37918d42c28033e9f684db52f94eeed617b

Changes since previous ACK:
* Avoided function overloading and documented that hostnames can be used in more places, not just raw IP addresses and the like.
* `--auto`
